src/ln/channelmonitor.rs

```diff
@@ -286,9 +286,25 @@ impl ChannelMonitor {
 				}
 			}
 		}
-		// TODO: Prune payment_preimages no longer needed by the revocation (just have to check
-		// that non-revoked remote commitment tx(n) do not need it, and our latest local commitment
-		// tx does not need it.
+
+		let local_signed_commitment_tx = &self.current_local_signed_commitment_tx;
+		let remote_claimable_outpoints = &self.remote_claimable_outpoints;
+		self.payment_preimages.retain(|&k, _| {
+				for &(ref htlc, _s1, _s2) in &local_signed_commitment_tx.as_ref().expect("Channel needs at least an initial commitment tx !").htlc_outputs {
+					if k == htlc.payment_hash {
+						return true
+					}
+				}
+				for (_tx, htlcs) in remote_claimable_outpoints {
+					for htlc in htlcs {
+						if k == htlc.payment_hash {
```

Comment:
Still kinda worried about the potential performance issue here, especially for watchtowers.